File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/notebook/NotebookQueue.kt
package com.riskdetectedan.core.data.notebook

import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.util.UUID

class NotebookFailure(val code: String) : Exception(code)
class NotebookServerFailure(val code: String) : Exception(code)
data class NotebookIdentity(val owner: UUID, val session: UUID)
@Serializable
data class NotebookMutation(val mutation: String, val note: String, val action: String, val expected: Long,
    val title: String?, val body: String?, val conflict: String?, val items: List<NotebookItem>? = null, val tags: List<String>? = null) {
    fun validate() {
        require(UUID.fromString(mutation).toString() == mutation && UUID.fromString(note).toString() == note)
        require(conflict == null || UUID.fromString(conflict).toString() == conflict)
        if (action == "organize") {
            require(expected in 1..9007199254740990L && title == null && body == null && conflict == null && items != null && tags != null)
            validateNotebookOrganization(items, tags); return
        }
        require(items == null && tags == null)
        require(action in setOf("sync", "delete", "resolve") && expected in 0..9007199254740990L)
        require((title?.codePointCount(0, title.length) ?: 0) <= 200 && (body?.codePointCount(0, body.length) ?: 0) <= 20000)
        require((action == "resolve") == (conflict != null))
        require(if (action == "delete") title == null && body == null else title != null || body != null)
    }
    fun arguments(): JsonObject {
        validate()
        if (action == "organize") return buildJsonObject {
            put("p_mutation", mutation); put("p_note", note); put("p_expected", expected)
            put("p_items", Json.encodeToJsonElement(items!!)); put("p_tags", Json.encodeToJsonElement(tags!!))
        }
        return buildJsonObject {
            put("p_mutation", mutation); put("p_note", note); put("p_action", action); put("p_expected", expected)
            put("p_title", title?.let(::JsonPrimitive) ?: JsonNull); put("p_body", body?.let(::JsonPrimitive) ?: JsonNull)
            put("p_conflict", conflict?.let(::JsonPrimitive) ?: JsonNull)
        }
    }
}
@Serializable
data class NotebookPending(val intent: NotebookMutation, val attempted: Boolean = false, val blocked: String? = null, val conflictID: String? = null)
interface NotebookStorage { fun read(owner: UUID): String?; fun write(owner: UUID, value: String) }

/** Explicitly driven durable outbox. No automatic launch, company dependency, or paid gate. */
class NotebookQueue(private val storage: NotebookStorage, private val current: () -> NotebookIdentity?,
    private val send: suspend (JsonObject) -> JsonObject) {
    @Serializable private data class Ledger(val schema: Int = 1, val owner: String, val entries: List<NotebookPending>)
    companion object { private val lock = Mutex() }
    private val json = Json { encodeDefaults = true; ignoreUnknownKeys = false }
    private suspend fun check(identity: NotebookIdentity) {
        currentCoroutineContext().ensureActive()
        if (current() != identity) throw NotebookFailure("IDENTITY_CHANGED")
    }
    private fun load(owner: UUID): Ledger {
        val value = storage.read(owner) ?: return Ledger(owner = owner.toString(), entries = emptyList())
        require(value.toByteArray(Charsets.UTF_8).size <= 2_000_000)
        val ledger = json.decodeFromString<Ledger>(value)
        require(ledger.schema == 1 && ledger.owner == owner.toString() && ledger.entries.size <= 20)
        require(ledger.entries.map { it.intent.mutation }.toSet().size == ledger.entries.size)
        ledger.entries.forEach { it.intent.validate() }; return ledger
    }
    private fun persist(ledger: Ledger) {
        val value = json.encodeToString(ledger)
        if (value.toByteArray(Charsets.UTF_8).size > 2_000_000) throw NotebookFailure("QUEUE_FULL")
        storage.write(UUID.fromString(ledger.owner), value)
    }
    suspend fun pending(identity: NotebookIdentity): List<NotebookPending> = lock.withLock { check(identity); load(identity.owner).entries }
    suspend fun stage(intent: NotebookMutation, identity: NotebookIdentity) = lock.withLock {
        check(identity); intent.validate(); val ledger = load(identity.owner)
        ledger.entries.find { it.intent.mutation == intent.mutation }?.let { require(it.intent == intent); return@withLock }
        if (ledger.entries.size >= 20) throw NotebookFailure("QUEUE_FULL")
        persist(ledger.copy(entries = ledger.entries + NotebookPending(intent)))
    }
    /** Explicit user recovery action, never automatically invoked on errors. */
    suspend fun discard(mutation: String, identity: NotebookIdentity) = lock.withLock {
        check(identity); val ledger = load(identity.owner)
        persist(ledger.copy(entries = ledger.entries.filterNot { it.intent.mutation == mutation }))
    }
    suspend fun resolveBlocked(mutation: String, replacement: NotebookMutation, identity: NotebookIdentity) = lock.withLock {
        check(identity); replacement.validate(); val ledger = load(identity.owner)
        val index = ledger.entries.indexOfFirst { it.intent.mutation == mutation }; require(index >= 0)
        val prior = ledger.entries[index]
        require(prior.blocked == "VERSION_CONFLICT" && prior.conflictID != null && replacement.action == "resolve" &&
            replacement.note == prior.intent.note && replacement.conflict == prior.conflictID && ledger.entries.none { it.intent.mutation == replacement.mutation })
        val entries = ledger.entries.toMutableList(); entries[index] = NotebookPending(replacement)
        persist(ledger.copy(entries = entries))
    }
    suspend fun syncNext(identity: NotebookIdentity): String = lock.withLock {
        check(identity); val ledger = load(identity.owner); val blockedNotes = mutableSetOf<String>()
        val index = ledger.entries.indexOfFirst {
            if (it.blocked != null) { blockedNotes += it.intent.note; false } else it.intent.note !in blockedNotes
        }
        if (index < 0) return@withLock "idle"
        val entries = ledger.entries.toMutableList(); entries[index] = entries[index].copy(attempted = true)
        persist(ledger.copy(entries = entries)); check(identity)
        val intent = entries[index].intent
        val ack = try { send(intent.arguments()) } catch (error: Exception) {
            check(identity)
            val code = (error as? NotebookServerFailure)?.code
            if (code in setOf("VERSION_CONFLICT", "NOTE_TOMBSTONED", "IDEMPOTENCY_CONFLICT", "ACCESS_DENIED", "VALIDATION_ERROR", "CONFLICT_ALREADY_RESOLVED")) {
                entries[index] = entries[index].copy(blocked = code,
                    conflictID = if (code == "VERSION_CONFLICT" && intent.action == "resolve") intent.conflict else entries[index].conflictID)
                persist(ledger.copy(entries = entries)); return@withLock "blocked"
            }
            throw NotebookFailure("UNAVAILABLE")
        }
        check(identity)
        require(ack.toString().toByteArray(Charsets.UTF_8).size <= 16384)
        require(ack["schema_version"]?.jsonPrimitive?.int == 1 && ack["mutation_id"]?.jsonPrimitive?.content == intent.mutation && ack["note_id"]?.jsonPrimitive?.content == intent.note)
        require(ack["replayed"]?.jsonPrimitive?.booleanOrNull != null)
        val state = ack["state"]?.jsonPrimitive?.content
        if (state == "conflict") {
            val conflict = ack["conflict_id"]!!.jsonPrimitive.content
            val version = ack["server_version"]!!.jsonPrimitive.long
            require(intent.action == "sync" && UUID.fromString(conflict).toString() == conflict && version in 1..9007199254740991L && ack["both_texts_preserved"]?.jsonPrimitive?.boolean == true)
            entries[index] = entries[index].copy(blocked = "VERSION_CONFLICT", conflictID = conflict)
            persist(ledger.copy(entries = entries)); return@withLock "conflict"
        }
        val allowed = when (intent.action) { "organize" -> setOf("organized"); "delete" -> setOf("deleted"); "resolve" -> setOf("resolved"); else -> setOf("created", "updated", "unchanged") }
        val version = ack["version"]!!.jsonPrimitive.long
        require(intent.action != "resolve" || ack["conflict_id"]?.jsonPrimitive?.content == intent.conflict)
        require(state in allowed && version in 1..9007199254740991L && (state == "deleted" || version == intent.expected + if (state == "unchanged") 0 else 1))
        entries.removeAt(index); persist(ledger.copy(entries = entries)); "committed"
    }
    suspend fun replaceOrganization(mutation: String, replacement: NotebookMutation, identity: NotebookIdentity) = lock.withLock {
        check(identity); replacement.validate(); val ledger = load(identity.owner)
        val index = ledger.entries.indexOfFirst { it.intent.mutation == mutation }; require(index >= 0)
        val prior = ledger.entries[index]
        require(prior.blocked == "VERSION_CONFLICT" && prior.intent.action == "organize" && replacement.action == "organize" &&
            replacement.note == prior.intent.note && ledger.entries.none { it.intent.mutation == replacement.mutation })
        val entries = ledger.entries.toMutableList(); entries[index] = NotebookPending(replacement); persist(ledger.copy(entries = entries))
    }
}

File: App/Services/Notebook/NotebookQueue.swift
import Foundation

enum NotebookFailure: String, Error { case invalid, unavailable, identityChanged, full, busy }
struct NotebookIdentity: Equatable { let owner: UUID; let session: UUID }
struct NotebookMutation: Codable, Equatable {
    let mutation: UUID
    let note: UUID
    let action: String
    let expected: Int64
    let title: String?
    let body: String?
    let conflict: UUID?
    var items: [NotebookItem]? = nil
    var tags: [String]? = nil
    func validate() throws {
        if action == "organize" {
            guard (1...9007199254740990).contains(expected), title == nil, body == nil, conflict == nil,
                  let items, let tags else { throw NotebookFailure.invalid }
            try validateNotebookOrganization(items, tags); return
        }
        guard items == nil, tags == nil else { throw NotebookFailure.invalid }
        guard ["sync", "delete", "resolve"].contains(action), (0...9007199254740990).contains(expected),
              (title?.unicodeScalars.count ?? 0) <= 200, (body?.unicodeScalars.count ?? 0) <= 20000,
              (action == "resolve") == (conflict != nil),
              action == "delete" ? title == nil && body == nil : title != nil || body != nil else { throw NotebookFailure.invalid }
    }
    func arguments() throws -> Data {
        try validate()
        if action == "organize" {
            let encoder = JSONEncoder(); encoder.outputFormatting = [.sortedKeys]
            return try encoder.encode(NotebookOrganizationPayload(p_mutation: mutation, p_note: note, p_expected: expected, p_items: items!, p_tags: tags!))
        }
        return try JSONSerialization.data(withJSONObject: ["p_mutation": mutation.uuidString.lowercased(),
            "p_note": note.uuidString.lowercased(), "p_action": action, "p_expected": expected,
            "p_title": title as Any? ?? NSNull(), "p_body": body as Any? ?? NSNull(),
            "p_conflict": conflict?.uuidString.lowercased() as Any? ?? NSNull()], options: [.sortedKeys])
    }
}
struct NotebookPending: Codable, Equatable {
    let intent: NotebookMutation
    var attempted: Bool = false
    var blocked: String?
    var conflictID: UUID?
}
@MainActor protocol NotebookStorage {
    func read(owner: UUID) throws -> Data?
    func write(_ data: Data, owner: UUID) throws
}
struct NotebookServerFailure: Error { let code: String }

/// One owner-bound durable outbox. No company scope, subscription check or telemetry.
/// Instantiation does not send. A caller must explicitly drive syncNext after rollout.
@MainActor final class NotebookQueue {
    typealias Send = (Data) async throws -> Data
    private struct Ledger: Codable { let schema: Int; let owner: UUID; var entries: [NotebookPending] }
    private static var sending = Set<UUID>()
    private let storage: any NotebookStorage
    private let current: () -> NotebookIdentity?
    private let send: Send
    init(storage: any NotebookStorage, current: @escaping () -> NotebookIdentity?, send: @escaping Send) {
        self.storage = storage; self.current = current; self.send = send
    }
    private func check(_ identity: NotebookIdentity) throws {
        try Task.checkCancellation()
        guard current() == identity else { throw NotebookFailure.identityChanged }
    }
    private func load(_ owner: UUID) throws -> Ledger {
        guard let data = try storage.read(owner: owner) else { return Ledger(schema: 1, owner: owner, entries: []) }
        guard data.count <= 2_000_000 else { throw NotebookFailure.invalid }
        let ledger = try JSONDecoder().decode(Ledger.self, from: data)
        guard ledger.schema == 1, ledger.owner == owner, ledger.entries.count <= 20,
              Set(ledger.entries.map { $0.intent.mutation }).count == ledger.entries.count else { throw NotebookFailure.invalid }
        for entry in ledger.entries { try entry.intent.validate() }
        return ledger
    }
    private func persist(_ ledger: Ledger) throws {
        let data = try JSONEncoder().encode(ledger)
        guard data.count <= 2_000_000 else { throw NotebookFailure.full }
        try storage.write(data, owner: ledger.owner)
    }
    func pending(_ identity: NotebookIdentity) throws -> [NotebookPending] {
        try check(identity); return try load(identity.owner).entries
    }
    func stage(_ intent: NotebookMutation, identity: NotebookIdentity) throws {
        try check(identity); try intent.validate()
        var ledger = try load(identity.owner)
        if let existing = ledger.entries.first(where: { $0.intent.mutation == intent.mutation }) {
            guard existing.intent == intent else { throw NotebookFailure.invalid }; return
        }
        guard ledger.entries.count < 20 else { throw NotebookFailure.full }
        ledger.entries.append(NotebookPending(intent: intent)); try persist(ledger)
    }
    // Explicit recovery action only; never called automatically on an RPC error.
    func discard(_ mutation: UUID, identity: NotebookIdentity) throws {
        try check(identity); guard !Self.sending.contains(identity.owner) else { throw NotebookFailure.busy }
        var ledger = try load(identity.owner)
        ledger.entries.removeAll { $0.intent.mutation == mutation }; try persist(ledger)
    }
    func resolveBlocked(_ mutation: UUID, with replacement: NotebookMutation, identity: NotebookIdentity) throws {
        try check(identity); try replacement.validate()
        guard !Self.sending.contains(identity.owner) else { throw NotebookFailure.busy }
        var ledger = try load(identity.owner)
        guard let i = ledger.entries.firstIndex(where: { $0.intent.mutation == mutation }),
              ledger.entries[i].blocked == "VERSION_CONFLICT", let conflict = ledger.entries[i].conflictID,
              replacement.action == "resolve", replacement.note == ledger.entries[i].intent.note, replacement.conflict == conflict,
              !ledger.entries.contains(where: { $0.intent.mutation == replacement.mutation }) else { throw NotebookFailure.invalid }
        ledger.entries[i] = NotebookPending(intent: replacement); try persist(ledger)
    }
    func replaceOrganization(_ mutation: UUID, with replacement: NotebookMutation, identity: NotebookIdentity) throws {
        try check(identity); try replacement.validate()
        guard !Self.sending.contains(identity.owner) else { throw NotebookFailure.busy }
        var ledger = try load(identity.owner)
        guard let i = ledger.entries.firstIndex(where: { $0.intent.mutation == mutation }),
              ledger.entries[i].blocked == "VERSION_CONFLICT", ledger.entries[i].intent.action == "organize",
              replacement.action == "organize", replacement.note == ledger.entries[i].intent.note,
              !ledger.entries.contains(where: { $0.intent.mutation == replacement.mutation }) else { throw NotebookFailure.invalid }
        ledger.entries[i] = NotebookPending(intent: replacement); try persist(ledger)
    }
    func syncNext(_ identity: NotebookIdentity) async throws -> String {
        try check(identity)
        guard !Self.sending.contains(identity.owner) else { throw NotebookFailure.busy }
        Self.sending.insert(identity.owner); defer { Self.sending.remove(identity.owner) }
        var ledger = try load(identity.owner)
        // A blocked edit must not be overtaken by a later edit to the same note.
        var blockedNotes = Set<UUID>()
        let index = ledger.entries.indices.first { i in
            let entry = ledger.entries[i]
            if entry.blocked != nil { blockedNotes.insert(entry.intent.note); return false }
            return !blockedNotes.contains(entry.intent.note)
        }
        guard let index else { return "idle" }
        ledger.entries[index].attempted = true
        let intent = ledger.entries[index].intent
        try persist(ledger) // Must succeed before network; same UUID survives restart.
        try check(identity)
        let data: Data
        do { data = try await send(intent.arguments()) }
        catch {
            try check(identity)
            if let code = (error as? NotebookServerFailure)?.code,
               ["VERSION_CONFLICT","NOTE_TOMBSTONED","IDEMPOTENCY_CONFLICT","ACCESS_DENIED","VALIDATION_ERROR","CONFLICT_ALREADY_RESOLVED"].contains(code) {
                var latest = try load(identity.owner)
                guard let i = latest.entries.firstIndex(where: { $0.intent == intent }) else { throw NotebookFailure.invalid }
                latest.entries[i].blocked = code
                if code == "VERSION_CONFLICT", intent.action == "resolve" { latest.entries[i].conflictID = intent.conflict }
                try persist(latest); return "blocked"
            }
            throw NotebookFailure.unavailable // Preserve the draft, never expose raw errors.
        }
        try check(identity)
        struct Receipt: Decodable {
            let schema_version: Int; let mutation_id: UUID; let note_id: UUID; let state: String
            let replayed: Bool; let version: Int64?; let server_version: Int64?; let conflict_id: UUID?; let both_texts_preserved: Bool?
        }
        guard data.count <= 16384 else { throw NotebookFailure.invalid }
        let ack = try JSONDecoder().decode(Receipt.self, from: data)
        guard ack.schema_version == 1, ack.mutation_id == intent.mutation, ack.note_id == intent.note else { throw NotebookFailure.invalid }
        var latest = try load(identity.owner)
        guard let i = latest.entries.firstIndex(where: { $0.intent == intent }) else { throw NotebookFailure.invalid }
        if ack.state == "conflict" {
            guard intent.action == "sync", let conflict = ack.conflict_id, let version = ack.server_version,
                  (1...9007199254740991).contains(version), ack.both_texts_preserved == true else { throw NotebookFailure.invalid }
            latest.entries[i].blocked = "VERSION_CONFLICT"; latest.entries[i].conflictID = conflict
            try persist(latest); return "conflict"
        }
        let allowed = intent.action == "organize" ? ["organized"] : intent.action == "delete" ? ["deleted"] : intent.action == "resolve" ? ["resolved"] : ["created","updated","unchanged"]
        guard allowed.contains(ack.state), let version = ack.version, (1...9007199254740991).contains(version),
              intent.action != "resolve" || ack.conflict_id == intent.conflict,
              ack.state == "deleted" || version == intent.expected + (ack.state == "unchanged" ? 0 : 1) else { throw NotebookFailure.invalid }
        latest.entries.remove(at: i); try persist(latest)
        return "committed"
    }
}

struct NotebookOrganizationPayload: Codable {
    let p_mutation: UUID; let p_note: UUID; let p_expected: Int64
    let p_items: [NotebookItem]; let p_tags: [String]
}
